File: TasteKit/tastekit/packages/core/trust/auditor.ts
import { TrustV1 } from '../schemas/trust.js';
import { BindingsV1 } from '../schemas/bindings.js';

/**
 * Trust Auditor
 * 
 * Audits trust policy and flags violations.
 */

export interface AuditViolation {
  type: 'fingerprint_mismatch' | 'unpinned_server' | 'unpinned_skill' | 'new_tool';
  severity: 'error' | 'warning';
  message: string;
  details?: any;
}

export interface AuditReport {
  timestamp: string;
  violations: AuditViolation[];
  passed: boolean;
}

export class TrustAuditor {
  audit(trust: TrustV1, bindings: BindingsV1): AuditReport {
    const violations: AuditViolation[] = [];
    
    // Check all bound servers are pinned
    for (const server of bindings.servers) {
      const pinned = trust.mcp_servers.find(s => s.url === server.url);
      
      if (!pinned) {
        violations.push({
          type: 'unpinned_server',
          severity: 'warning',
          message: `MCP server not pinned: ${server.name}`,
          details: { url: server.url },
        });
      } else if (server.pinned_fingerprint && server.pinned_fingerprint !== pinned.fingerprint) {
        violations.push({
          type: 'fingerprint_mismatch',
          severity: pinned.pin_mode === 'strict' ? 'error' : 'warning',
          message: `Fingerprint mismatch for server: ${server.name}`,
          details: {
            expected: pinned.fingerprint,
            actual: server.pinned_fingerprint,
          },
        });
      }
    }
    
    return {
      timestamp: new Date().toISOString(),
      violations,
      passed: violations.filter(v => v.severity === 'error').length === 0,
    };
  }
}

File: AutoClaw/internal/mcpclient/auditor.go
package mcpclient

import (
	"fmt"

	"github.com/philipbankier/autoclaw/internal/artifact"
)

// AuditFinding represents a single trust audit result.
type AuditFinding struct {
	ServerName string `json:"server_name"`
	Severity   string `json:"severity"` // "error"|"warning"|"info"
	Message    string `json:"message"`
}

// Auditor cross-references MCP trust policy (from TasteKit trust artifact)
// against the active server registry and trust pins.
type Auditor struct {
	registry *Registry
	trust    *TrustManager
}

// NewAuditor creates an auditor from a registry and trust manager.
func NewAuditor(registry *Registry, trust *TrustManager) *Auditor {
	return &Auditor{registry: registry, trust: trust}
}

// Audit checks the current state against TasteKit trust policy.
func (a *Auditor) Audit(trustArtifact *artifact.TrustV1) []AuditFinding {
	var findings []AuditFinding

	// Check that all trust-pinned MCP servers are in the registry.
	registeredServers := a.registry.List()
	for _, srv := range trustArtifact.MCPServers {
		found := false
		for _, cfg := range registeredServers {
			// Match by command or URL-like reference.
			if cfg.Command == srv.URL || matchesServerURL(cfg, srv.URL) {
				found = true
				break
			}
		}
		if !found {
			findings = append(findings, AuditFinding{
				ServerName: srv.URL,
				Severity:   "warning",
				Message:    fmt.Sprintf("trust artifact references server %q but it is not in the registry", srv.URL),
			})
		}
	}

	// Check that all registered servers have trust pins.
	for name := range registeredServers {
		if _, ok := a.trust.Get(name); !ok {
			findings = append(findings, AuditFinding{
				ServerName: name,
				Severity:   "warning",
				Message:    fmt.Sprintf("registered server %q has no trust pin", name),
			})
		}
	}

	// Check that strict-mode pins have matching fingerprints.
	for name, pin := range a.trust.ListPins() {
		if pin.PinMode != "strict" {
			continue
		}
		if pin.Fingerprint == "" {
			findings = append(findings, AuditFinding{
				ServerName: name,
				Severity:   "error",
				Message:    fmt.Sprintf("server %q has strict pin mode but empty fingerprint", name),
			})
		}
	}

	if len(findings) == 0 {
		findings = append(findings, AuditFinding{
			Severity: "info",
			Message:  "all servers pass trust audit",
		})
	}

	return findings
}

func matchesServerURL(cfg ServerConfig, url string) bool {
	// Simple heuristic: check if the command contains the URL scheme or name.
	return false
}
